### PR TITLE
Upgrade to Jasmine 2.2.0

### PR DESCRIPTION
**Pull request by kring**
"Guaranteed to be the most fun you'll have reviewing a pull request all year!"

Lots of changes here:
* Remove `waitsFor`, `waits`, and `runs` from the list of globals in every spec file.  In about 200 of the 347 files changed, this is the only change.
* Use promises for async tests rather than `waitsFor` and `runs`, since Jasmine no longer supports those functions and promises are way better.
* Replace `toNotBe` and `toNotEqual` with `not.toBe` and `not.toEqual` because Jasmine no longer supports the former.
* Change every test that used spies, because the syntax changed a bit.
* A few small changes that were necessary because Jasmine's `toEqual` matcher is a lot pickier than it was before.  Not only must the properties of the objects be the same for the objects to be considered equal, but their constructors must be the same as well.
* In a few places, we were using `it` calls to set up preconditions for specs that followed for some reason.  This created order-dependence and made the dependent specs impossible to run by themselves.  I changed them to use `beforeAll` instead.
* Completely reimplement our custom features in the spec runner page, since merging was impossible.

On the last one, what I believe to be all the important functionality is working, such as coverage reporting, categories, and running against the built version, but I haven't done an amazing job of it.  The UI is a little clunkier than before.  I'd like to do more work on this but as a separate pull request so I'm not constantly chasing changes in Cesium's tests in the meantime.  If there's something you can't live with/without, let me know.

FAQ:
> Do I really have to review this?

No, but you should probably at least try running the specs in this branch to see how mad you'll be when it gets merged.

> Are we really sticking with Jasmine instead of using something else?

Maybe.  I started on this pull request in part to validate what I was writing in WebGL Insights.  It turned out to be a LOT more work than expected and I possibly should have taken better stock of other alternatives somewhere along the way.  The good news is, the changes to all the async tests to use promises are valuable pretty much no matter what test framework we use.

> What about Karma?  Upgrading to JSCover?  Implementing my favorite feature?

Maybe, but let's merge this first so my life sucks less. :)

### CODE REVIEW

**Comment by pjcozzi** [comment]
Thanks @kring.

@mramato how will this impact KML for 1.7?  Should we wait to merge this until after 1.7?  Or would you rather get the KML tests upgraded right away?  Or perhaps @kring will upgrade them for you...

**Comment by mramato** [comment]
@shunter I know you don't have the time for a review, but do you have any thoughts on this PR?

@pjcozzi I would rather merge this sooner than later.  Updating any KML tests will be trivial and since this PR is a test-only change, there's no chance it will break Cesium.  I'm going to do a review now, but I don't expect to have a lot of comments.

**Comment by mramato** [comment]
Thanks again @kring for taking point on this. I ran through all of the different testing configurations and everything seems to work as expected.

I'm fine with taking the UI as-is for now, but as you already noted this is clearly a step backward in that department.  I think the two main things I would want back ASAP are the debug and coverage links that would show up to the right of each test. I personally don't care whether we put back the show-as-you-go non-expanded table thing we had before.

I'm getting a bunch of `Expected [object Uint8Array] to equal [ 0, 0, 0, 255 ].` failures under Firefox. I think these tests were already failing in master but there it prints out the contents of the array rather than `[object Uint8Array]`.  I assume this would be an easy change?

Other than the above comments (both of which could be done post-merge) that's all I have. Assuming no one else wants to take a look at this, @pjcozzi feel free to merge this whenever you want.  Like I said in my initial comment, the sooner the better because it's going to be hell to maintain this branch.

**Comment by mramato** [comment]
FYI, to save others some time, I ran `ant makeZipFile` and ran all tests to make sure everything still works with the zip package (which it does).

**Comment by pjcozzi** [comment]
Looking at this now.  Will merge today if possible.

**Comment by pjcozzi** [comment]
Did you test in IE?  I don't get this failure in master:

image

File: Specs/Renderer/loadCubeMapSpec.js [hunk 1]
```diff
@@ -109,117 +100,93 @@ defineSuite([
...
             positiveY : './Data/Images/Blue.png',
             negativeY : './Data/Images/Blue.png',
             positiveZ : './Data/Images/Blue.png',
             negativeZ : './Data/Images/Blue.png'
-        }), function(cubeMap) {
-        }, function() {
-            exception = true;
+        }).then(function(cubeMap) {
+            // this should not be called.
+            expect(true).toBe(false);
```
Comment (pjcozzi) [inline Specs/Renderer/loadCubeMapSpec.js]: No action required, but I'm surprised they don't have basically `assert(false);`

File: Specs/Scene/BillboardCollectionSpec.js [hunk 2]
```diff
@@ -74,21 +80,19 @@ defineSuite([
...
-            whiteImage = new Image();
-            whiteImage.src = './Data/Images/White.png';
-
-            largeBlueImage = new Image();
-            largeBlueImage.src = './Data/Images/Blue10x10.png';
-
-            waitsFor(function() {
-                return greenImage.complete && blueImage.complete && whiteImage.complete && largeBlueImage.complete;
-            }, 'Load .png file(s) for billboard collection test.', 3000);
+            return when.join(
```
Comment (pjcozzi) [inline Specs/Scene/BillboardCollectionSpec.js]: Putting this in `beforeEach` is a workaround.  It should go in `beforeAll` without `if (!greenImage) {`.  Do you think that will work now in Jasmine 2.0?

File: Specs/Scene/MultifrustumSpec.js [hunk 3]
```diff
@@ -69,18 +73,16 @@ defineSuite([
...
-        blueImage = new Image();
-        blueImage.src = './Data/Images/Blue.png';
-
-        whiteImage = new Image();
-        whiteImage.src = './Data/Images/White.png';
-
-        waitsFor(function() {
-            return greenImage.complete && blueImage.complete && whiteImage.complete;
-        }, 'Load .png file(s) for billboard collection test.', 3000);
+        return when.join(
```
Comment (pjcozzi) [inline Specs/Scene/MultifrustumSpec.js]: Same question.

File: Specs/Scene/TextureAtlasSpec.js [hunk 4]
```diff
@@ -47,27 +47,25 @@ defineSuite([
...
-            bigBlueImage = new Image();
-            bigBlueImage.src = './Data/Images/Blue10x10.png';
-
-            bigGreenImage = new Image();
-            bigGreenImage.src = './Data/Images/Green4x4.png';
-
-            waitsFor(function() {
-                return greenImage.complete && tallGreenImage.complete && blueImage.complete && bigRedImage.complete && bigBlueImage.complete && bigGreenImage.complete;
-            }, 'Load .png file(s) for texture atlas test.', 3000);
+            return when.join(
```
Comment (pjcozzi) [inline Specs/Scene/TextureAtlasSpec.js]: Same question.

**Comment by pjcozzi** [comment]
Can you update LICENSE.md?  Probably CHANGES.md too.

**Comment by pjcozzi** [comment]
Across Chrome, FF, and IE, the tests appear to run faster.

**Comment by pjcozzi** [comment]
"Select a test to run" is also way faster.

**Comment by pjcozzi** [comment]
> I think the two main things I would want back ASAP are the debug and coverage links that would show up to the right of each test.

+1.  How hard is the debug link?

Given the links in index.html, I don't think we need a replacement for this:

image

> I'm getting a bunch of Expected [object Uint8Array] to equal [ 0, 0, 0, 255 ]. failures under Firefox. I think these tests were already failing in master but there it prints out the contents of the array rather than [object Uint8Array]. I assume this would be an easy change?

Yes, already failing in master.  I agree we'll want to actual values printed.

**Comment by pjcozzi** [comment]
I get this failure _sometimes_ when testing against the combined file:

image

Is it related to this upgrade?

**Comment by pjcozzi** [comment]
That's all.  Overall, looks awesome.  Thanks for this!

**Comment by kring** [comment]
> I think the two main things I would want back ASAP are the debug and coverage links that would show up to the right of each test.

Interesting, I don't think I've ever used those.  It's probably not hard to add them back, though, so I'll look at that once I've addressed everything else that strictly keeps this from being merged.

**Comment by mramato** [comment]
> Interesting, I don't think I've ever used those. It's probably not hard to add them back, though, so I'll look at that once I've addressed everything else that strictly keeps this from being merged.

I find debug useful when I don't have the code open in dev tools already and coverage is useful when I'm writing new unit tests and want to make sure they cover what I expect them to cover. (and it's way faster to run coverage on a single suite).  But like I said, no rush.

**Comment by kring** [comment]
> since this PR is a test-only change, there's no chance it will break Cesium

There is one change to `PrimitivePipeline`:
https://github.com/AnalyticalGraphicsInc/cesium/pull/2495/files#diff-57817109445325435f3dbaf33939cad0

This was in response to Jasmine's new pickiness about matching constructor functions in `toEqual`, but I think it's a good change anyway.

**Comment by kring** [comment]
> I get this failure sometimes when testing against the combined file:

I'm surprised you're only getting that sometimes.  I get it every time in the built version, and in master, too.  I think I can fix it easily enough though.

**Comment by kring** [comment]
Well fixing `Scene/Primitive failed geometry rejects promise and throws on next update` was kind of a hassle, and my fix is a hack, but it's done.  The problem is that the test requires that both the standard Cesium `createGeometry` and the spec-local `createBadGeometry` be loadable.  Since these are in completely different places, and their relative paths are different in built versus non-built Cesium, tricking the whole AMD tower into actually being able to load both of them is challenging.

**Comment by mramato** [comment]
You can blame me for BadGeometry.  I'm not sure there's a better way to do it and no one had any suggestions in the PR that introduced it.

**Comment by kring** [comment]
> Putting this in beforeEach is a workaround. It should go in beforeAll without if (!greenImage) {. Do you think that will work now in Jasmine 2.0?

Seems to work.  I can't think of why it wouldn't have worked in 1.3.

**Comment by kring** [comment]
> Did you test in IE? I don't get this failure in master:

I'm not getting anywhere quickly on this one.  I also see the failure, and don't see it in master.  It doesn't happen when I run the test by itself.  The test is a little sketchy in that it tries to set the camera position/orientation, but that doesn't actually work in our tests.  Even if I remove that code, the tests still pass in other browsers but not IE.  Presumably it's an order or timing problem.

It's going to take awhile to debug because it only happens when running all tests, and for reasons that are unclear IE11 is only able to run the test suite without hundreds of failures maybe one out of five times on my system.  The other 4/5 times, context creation starts failing. midway through.

So, any ideas welcome!

**Comment by pjcozzi** [comment]
Given that the `renders` test always passes, I'm surprised this fails.  What if you move this one above `renders`?

Perhaps this [epsilon workaround](https://github.com/AnalyticalGraphicsInc/cesium/blob/master/Source/Scene/DebugModelMatrixPrimitive.js#L129) is not big enough?

Perhaps we are better off just using a spy to see if a pick command was generated since picking tests like this, in general, can be a bit flakey.

> The test is a little sketchy in that it tries to set the camera position/orientation, but that doesn't actually work in our tests. 

Why not?  I thought we did this or something similar in lots of tests?

File: Specs/Renderer/loadCubeMapSpec.js (hunk 1 above)
Comment (shunter) [inline Specs/Renderer/loadCubeMapSpec.js]: They do.  

```javascript
fail('should not be called');
```

**Comment by shunter** [comment]
Minor, but I believe the previous default timeout on `waitsFor` was 5000ms.  The new `pollToPromise` uses 4000ms.

File: Specs/DataSources/CzmlDataSourceSpec.js [hunk 5]
```diff
@@ -270,11 +266,9 @@ defineSuite([
     });
 
     it('process loads data on top of existing', function() {
-        waitsFor(function() {
+        return pollToPromise(function() {
             return defined(simple) && defined(vehicle);
```
Comment (shunter) [inline Specs/DataSources/CzmlDataSourceSpec.js]: Assuming that the changes to `beforeAll` above work properly (that is, assuming beforeAll doesn't complete until the returned promise has resolved), I think this `pollToPromise` here (and similar throughout) is unnecessary.

**Comment by shunter** [comment]
> The problem is that the test requires that both the standard Cesium createGeometry and the spec-local createBadGeometry be loadable.

The better way to deal with this would be to modify `createGeometry` to also receive loader config information from the main environment into the worker environment, similar to how TaskProcessor does.  Probably baseUrl would be sufficient, so the module ID would become non-relative.  Then the test can override that information at test time to calculate the proper path that `createGeometry` would use to load the module out of the test directory.

That said, those changes are probably somewhat off-topic for this PR.

**Comment by shunter** [comment]
Other comments that I can't comment directly on because GitHub sucks at large pull requests:

> SpecRunner.html

We should keep the CSS for the Open Sans font.  We use that font in some of the text rendering tests to get a consistent font appearance across platforms.  You need both the CSS that defines the web font and the span that makes it actually load the font, because browsers suck.

**Comment by pjcozzi** [comment]
Any update here?  Do we expect to merge this before 1.7?  If so, it would be nice to merge soon so we can update the kml branch.

**Comment by kring** [comment]
Sorry, we had a big trade show thing Thursday and Friday last week, so I had no time to work on it.  I'll get back to it Monday morning so with any luck it will be ready before the weekend is over in your time zone.

**Comment by kring** [comment]
> > The test is a little sketchy in that it tries to set the camera position/orientation, but that doesn't actually work in our tests.
> 
> Why not? I thought we did this or something similar in lots of tests?

Sorry, what I said before is probably not true, because this test actually creates a `Scene`.  We have other tests (such as `RectanglePrimitiveSpec` that do rendering but skip the scene and manually do `createFrameState`.  With that approach, I believe setting camera properties won't work without explicitly updating the camera afterward.

**Comment by kring** [comment]
This should be ready.  I haven't brought back the Debug and Coverage links, and I haven't made Scott's suggested improvement to `createGeometry`.  I'll look at both of those separately.

re: that IE11 failure, I'm not seeing it anymore.  Maybe it was fixed with one of the changes I made at Scott's suggestion?  My IE 11 test runs are still really unstable though, so I don't know for sure.

**Comment by pjcozzi** [comment]
@mramato or @shunter can merge this when ready.

**Comment by mramato** [comment]
I hate to say it, but IE seems to be totally busted,  The Primitive Culling tests cause the browser to grind away on the CPU forever (not an infinite loop, just starts going really slow and failing).  After 5 minutes they still weren't done.  master doesn't have this problem.  I'm looking into it, but this isn't an area of the code I'm familiar with either.

I definitely want to get this merged soon, but if we do need to keep it on a branch a little longer, I have no problem helping keep it up to date with merges from master.

**Comment by mramato** [comment]
I also noticed some code coverage differences between master and this branch (which is at 92% coverage instead of 93).  I compared them side-by-side and as best I can tell it looks like there's 2 reasons.

1. Improved promise support means some tests are no longer polling, which isn't exercising functionality by accident like it used to.
2. Code in `beforeAll` does not affect coverage results.

Neither of these concern me regarding bringing this into master, I just wanted to mention it.  I'll try make sure we capture all of this stuff into an issue once this goes into master.

**Comment by kring** [comment]
> I hate to say it, but IE seems to be totally busted

For me, this is the case in both this branch and master, about 4/5 times I run the tests.  Are you sure you didn't just get lucky with your master run?

**Comment by kring** [comment]
> Code in beforeAll does not affect coverage results.

Weird, I can't think of why that would be, but I'll check it out.

**Comment by mramato** [comment]
> For me, this is the case in both this branch and master, about 4/5 times I run the tests. Are you sure you didn't just get lucky with your master run?

I don't know, I ran it again twice in master without problems, then I ran it 3 times in this branch and it failed twice.  It looks like either a race condition or some sort of context loss issue to me; but IE makes it really hard to track down and it only seems to happen when running the whole suite.

> Weird, I can't think of why that would be, but I'll check it out.

My assumption may be incorrect, but I remember this always being the case.  @shunter may know for sure.  It's possible that I'm mistaken.  Either way, the nature of the coverage loss wasn't anything to worry about in the short term (and we can just add tests back to explicitly cover stuff that we might be missing).

**Comment by kring** [comment]
Ok I just submitted a partial fix for the IE problems.  I had a bug that resulted in an exception escaping a spec function waiting for the timeout to expire instead of failing fast.  With this fix, there's no more hang, but there's still hundreds of test failures when IE11 loses the ability to create WebGL contexts for some reason.

**Comment by kring** [comment]
Here's what IE spits out in the console right before specs start failing:

> WEBGL11257: WebGL content is taking too long to render on your GPU. Temporarily switching to software rendering.

Still not sure why this would only show up in this branch, but getting closer.

**Comment by mramato** [comment]
Can you try commenting out both `fails to generate mipmaps` in CubeMapSpec.js on your machine, that seems to fix the problem for me and I've just run the test twice in a row without problems.  If that's the case for you too, we can detect IE and skip those tests (and write an issue so we look into it in the future).

**Comment by mramato** [comment]
Whoops, I spoke to soon, it just died again.  Any idea why there would be "pending specs" at the end of the tests, here's what I'm seeing: `6388 specs, 927 failures, 2 pending specs`

**Comment by kring** [comment]
>  Any idea why there would be "pending specs" at the end of the tests

No, sorry, I don't think I've seen that.  Is this after my most recent push?

**Comment by mramato** [comment]
> No, sorry, I don't think I've seen that. Is this after my most recent push?

I had a couple of tests excluded with `xit`, I think that's what caused it.

**Comment by mramato** [comment]
FYI, I'm updating this branch to master.

**Comment by mramato** [comment]
So I've actually started to see a few occasional crashes in IE 11 in master too, not every time like in this branch; but it does happen.  My thinking would be to merge this right after 1.7 ships and then that gives us a month to iron out any issues that might pop up (and hopefully resolve the IE issue as well).  Keeping this branch around until IE is full working isn't reasonable because a)it's most likely a bug in IE and b) keeping this branch up to date is a nightmare.

**Comment by pjcozzi** [comment]
+1

**Comment by kring** [comment]
Yeah, that sounds reasonable to me.  I've still been working on this (the IE11 problems) when I have a chance.  Here are some fun facts:

* On my Windows 10 machine (IE Version: 11.0.9800.0, Update Versions: 11.0.8), the tests run to completion every time without context creation failing, and there are 14 failures.
* On my Windows 8.1 machine (IE Version: 11.0.9600.17631, Update Versions: 11.0.16), IE decides to switch to software rendering midway through the tests and then context creation starts failing.
* I have no idea which of those two IE versions above is the newer one. :)
* I sometimes (maybe 1/10 times) see IE print a console message that it is dropping to software rendering when I run the `PrimitiveCollection` tests alone, but it happens after all the tests have finished running.
* Removing the `PrimitiveCollection` tests entirely makes the switch to software happen later, but it still happens.

I'm pretty much at a loss.  It's hard to see how it could be our problem (as opposed to an IE bug), but who knows?

**Comment by mramato** [comment]
> On my Windows 10 machine

Out of curiosity, have you tried Spartan yet?

**Comment by kring** [comment]
There's no actual Spartan browser in current Windows 10 previews, but you can turn on the new engine within Internet Explorer.  I did that, and I now see 56 test failures on this machine (versus 14 with the Trident engine), many of the failures in `ScreenSpaceCameraController` strangely enough.  I tried out a few Sandcastle examples and everything I tried was working well.  I would describe the performance as "not noticeably different from Chrome", but that's not very scientific.

**Comment by mramato** [comment]
> many of the failures in ScreenSpaceCameraController strangely enough

This doesn't surprise me, the whole touch/pointer events fiasco is most likely to blame.  We need to improve how we detect pointer events and shut off touch.  It would be great if we could use a polyfill, but that's not really an option for a library like Cesium.

**Comment by mramato** [comment]
Also, I merged in master again and updated all of the KML tests to work.  Once 1.7 is out tomorrow, it should be easy to do one last merge and then pull this into master.

**Comment by mramato** [comment]
We may be doing a 1.7.1 release, so I'll hold off on merging this into master until that happens.

**Comment by mramato** [comment]
I merged in master, but this branch is currently broken because ModelSpec had a lot of changes and there were some other stray failures.  I plan on fixing them and merging this into master hopefully sometime on Monday (EST)

**Comment by kring** [comment]
I fixed the model tests by reverting to the version in master and then re-upgrading it for Jasmine 2.2.  There are still 91 failures though (all in geometries I think), which I'll look at soon if you don't beat me to it.

**Comment by mramato** [comment]
Thank @kring.  Those are related to the generated packing specs and are failing because we are comparing a plain object to an instance with the same properties (which as you mentioned earlier in the PR, fail in Jasmine 2.  Should be an easy fix.

**Comment by mramato** [comment]
Alright, tests are all passing.  To be on the safe side, I ran the build a final time and tested with the release zip as well.  As far as I'm concerned we are good to go.  Thanks again for taking this up @kring, while the promise based tests are a good win, I'm more excited for karma and other benefits we'll hopefully see in the near future.

**Pull request merged**